### PR TITLE
[feature](Nereids) implement grouping sets

### LINKED ISSUE #13615: support grouping sets in Nereids
The current new optimizer does not support related keywords such as grouping sets, as well as grouping and grouping_id functions. The purpose of this issue is to support related functions and syntax.
The related syntax and functions are used in the same way as the old optimizer, refer to:
https://doris.apache.org/zh-CN/community/design/grouping_sets_design

Subtasks for the issue:
- [x] Support group by grouping sets syntax #12583
- [ ] Support grouping and grouping_id functions #12583
- [ ] Support group by rollup and cube syntax
- [ ] Support group by rollup cube syntax

### PR DESCRIPTION
**Pull request by zhengshiJ**
# Proposed changes

Issue Number: close #13615 

## Problem summary

The pr implements the function:
1. Added parsing of grouping sets, rollup and cube.
2. grouping sets that implement the new optimizer.
3. Implement grouping and grouping_id functions.

**TODO**:

1. Implementation of cube and rollup.
2. Scenarios for groupBy with predicate pushdown added.

**Implementation details:**
Add the abstract class of LogicalGroupBy in the new optimizer, and the subsequent LogicalGroupingSets, LogicalGroupCube and LogicalRollUp will inherit from this abstract class.

Add the LogicalGroupBy operator between aggregate and scan, and finally convert the operator into RepeatNode in PhysicalPlanTranslate.

**Eg:**
```sql
select sum(k2), grouping(k1) from t1 group by grouping sets((k1));
```
parse:
```
LogicalProject 
->
LogicalAggregate 
(outputExpressions: sum(k2), grouping(k1); groupByExpressions: empty)
 ->
LogicalGroupBy 
(outputExpressions: sum(k2), grouping(k1); groupingSets: k1; groupByExpressions: k1)
-> 
LogicalScan
```

analyze:
```
LogicalProject 
->
LogicalAggregate 
(outputExpressions: sum(k2), grouping(k1); groupByExpressions: k1, GROUPING_ID(), GROUPING_PREFIX(k1))
 ->
LogicalGroupBy 
(outputExpressions: sum(k2), grouping(k1); groupingSets: k1; groupByExpressions: k1, GROUPING_ID(), GROUPING_PREFIX(k1))
-> 
LogicalScan
```

rewrite:
```
LogicalProject 
->
LogicalAggregate 
(outputExpressions: sum(k2), grouping(k1); groupByExpressions: k1, GROUPING_ID(), GROUPING_PREFIX(k1))
 ->
LogicalProject
(projections: k1, k2, GROUPING_ID(), GROUPING_PREFIX(k1))
->
LogicalGroupBy 
(outputExpressions: k1, k2, GROUPING_ID(), GROUPING_PREFIX(k1); groupingSets: k1; groupByExpressions: k1, GROUPING_ID(), GROUPING_PREFIX(k1))
-> 
LogicalScan
```
translate:
```
 * The function of toLongValues is to display the virtual value corresponding to virtualSlot.
 * For extra generated columns:
 *  Rule: The column corresponding to groupingIdList is a non-null column,
 *        and the value is 0, and the rest are null columns, and the value is 1.
 *
 * For groupingFunc columns:
 *  Rules:
 *      grouping: Only one parameter is allowed,
 *                when the column is an aggregation column, it is set to 1, otherwise it is 0.
 *      grouping_id: Multiple columns are allowed,
 *                   and the decimal result is returned based on the bitmap of the aggregated column.
 *
 * eg:
 * select k1, grouping(k1), grouping(k1, k2) from t1 group by grouping sets ((k1), (k1, k2), (k3) ());
 * nonVirtualGroupingExprs: [k1, k2, k3]
 * virtualSlotRefs: GROUPING_ID(), GROUPING_PREFIX_K1(k1), GROUPING_PREFIX_K1_K2(k1, k2)
 * bitSetAll: {0, 1, 2}
 * groupingIdList: [{0}, {0, 1}, {2}, {}]
 *
 * GROUPING_ID():
 * For {0}: k1 is set to 0, k2 and k3 are set to 1.
 * +----+----+----+
 * | k1 | k2 | k3 |
 * +----+----+----+
 * | 0  | 1  | 1  |
 * +----+----+----+
 * convert binary to decimal, the result is 3.
 *
 * And so on, The corresponding list of GROUPING_ID() is [3, 1, 6, 7]
 *
 * GROUPING_PREFIX_K1(k1):
 * For {0}:
 * +----+
 * | k1 |
 * +----+
 * | 0  |
 * +----+
 * And so on, The corresponding list GROUPING_PREFIX_K1(k1) is [0, 0, 1, 1]
 *
 * GROUPING_PREFIX_K1_K2(k1, k2):
 * For {0}:
 * +----+----+
 * | k1 | k2 |
 * +----+----+
 * | 0  | 1  |
 * +----+----+
 * convert binary to decimal, the result is 1.
 *
 * And so on, The corresponding list GROUPING_PREFIX_K1_K2(k1, k2) is [1, 0, 3, 3]
```

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
3. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
5. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
6. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/ChildOutputPropertyDeriver.java [hunk 1]
```diff
@@ -87,6 +89,18 @@ public PhysicalProperties visitPhysicalAggregate(PhysicalAggregate<? extends Pla
         }
     }
 
+    @Override
+    public PhysicalProperties visitPhysicalGroupBy(PhysicalGroupBy<? extends Plan> groupBy, PlanContext context) {
```
Comment (zhengshiJ) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/ChildOutputPropertyDeriver.java:93]: @morrySnow Please help to see if this is correct

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java [hunk 2]
```diff
@@ -113,6 +114,12 @@ public Void visitPhysicalAggregate(PhysicalAggregate<? extends Plan> agg, PlanCo
         // TODO: add other phase logical when we support distinct aggregate
     }
 
+    @Override
+    public Void visitPhysicalGroupBy(PhysicalGroupBy<? extends Plan> groupBy, PlanContext context) {
```
Comment (zhengshiJ) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java:118]: @morrySnow Please help to see if this is correct

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.53 seconds
 load time: 574 seconds
 storage size: 17154655310 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221027110141_clickbench_pr_34751.html

**Review by 924060929 — COMMENTED** [review]
The code is too complex, this is risk and may be have many potential bugs.

A possible solution is use LogicalAggregate in optimize stage, and generate GroupingFunctionCallExpr at translate stage

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalGroupingSets.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                groupingIdList.add(bitSet);
+            }
+        });
+        return groupingIdList;
+    }
+
+    public static List<Expression> genOriginalGroupByExpressions(List<List<Expression>> groupingSets) {
+        List<Expression> groupingExpression = new ArrayList<>();
+        groupingSets.forEach(groupingExpression::addAll);
+        return groupingExpression.stream().distinct().collect(ImmutableList.toImmutableList());
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalGroupingSets.java:225]:
```suggestion
        return groupingSets.stream()
                .flatMap(List::stream)
                .distinct()
                .collect(ImmutableList.toImmutableList());
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalGroupingSets.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                bitSet.set(i, expressions.contains(groupingExprs.get(i)));
+            }
+            if (!groupingIdList.contains(bitSet)) {
+                groupingIdList.add(bitSet);
+            }
+        });
+        return groupingIdList;
+    }
+
+    public static List<Expression> genOriginalGroupByExpressions(List<List<Expression>> groupingSets) {
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalGroupingSets.java:222]: `genOriginalGroupByExpressions`  -> `extractDistinctGroupByExpressions` ?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRepeat.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    public static final String GROUPING_PREFIX = "GROUPING_PREFIX_";
+    // max num of distinct sets in grouping sets clause
+    public static final int MAX_GROUPING_SETS_NUM = 64;
+    protected final List<Expression> groupByExpressions;
+    protected final List<NamedExpression> outputExpressions;
+    protected final Set<VirtualSlotReference> virtualSlotRefs;
+    protected final List<BitSet> groupingIdList;
+    protected final List<Expression> virtualGroupingExprs;
+    protected final boolean changedOutput;
+    protected final boolean isNormalized;
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRepeat.java:58]: We should have some comments for these members.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRepeat.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+/**
+ * PhysicalRepeat.
+ */
+public abstract class PhysicalRepeat<CHILD_TYPE extends Plan>
+        extends PhysicalUnary<CHILD_TYPE> implements Repeat {
+    protected final List<Expression> groupByExpressions;
+    protected final List<Expression> originGroupByExpressions;
+    protected final List<NamedExpression> outputExpressions;
+    protected final List<BitSet> groupingIdList;
+    protected final List<Expression> virtualGroupByExpressions;
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRepeat.java:48]: We should add some comments for these members. In particular,  `groupByExpressions` and `originGroupByExpressions` and confusing.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java [hunk 7]
```diff
@@ -380,4 +409,12 @@ private static org.apache.doris.analysis.AssertNumRowsElement.Assertion translat
...
         }
     }
+
+    public static GroupByClause.GroupingType translateGroupingType(
+            PhysicalRepeat repeat) {
+        if (repeat instanceof PhysicalGroupingSets) {
+            return GroupingType.GROUPING_SETS;
+        }
+        throw new AnalysisException("UnSupported type");
+    }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java:419]: PhysicalRepeat is a Plan, don't translate in ExpressionTranslator.
You can move to PhysicalPlanTranslator.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 8]
```diff
@@ -133,6 +140,26 @@ public PlanFragment translatePlan(PhysicalPlan physicalPlan, PlanTranslatorConte
...
+    private List<SlotReference> genGroupBySlotList(List<Expression> groupByExpressionList,
+            List<NamedExpression> outputExpressionList) {
+        List<SlotReference> groupSlotList = Lists.newArrayList();
+        Set<VirtualSlotReference> virtualSlotReferences = groupByExpressionList.stream()
+                .filter(VirtualSlotReference.class::isInstance)
+                .map(VirtualSlotReference.class::cast)
+                .collect(Collectors.toSet());
+        for (Expression e : groupByExpressionList) {
+            if (e instanceof SlotReference && outputExpressionList.stream().anyMatch(o -> o.anyMatch(e::equals))) {
+                groupSlotList.add((SlotReference) e);
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:154]:
```suggestion
            if (virtualSlotReferences.contains(e))) {
                groupSlotList.add((VirtualSlotReference) e);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 9]
```diff
@@ -257,6 +283,193 @@ public PlanFragment visitPhysicalAggregate(
...
+     * 1. Put the columns in groupBy into the slot
+     * 2. Put the columns in the output into the slot
+     * 3. Put slot and virtualSlot into finalSLots in turn.
+     *
+     * eg: select sum(k2) grouping(k1) from t1 group by grouping sets((k1));
+     * 1. slots:[k1]
+     * 2. slots:[k1, k2]
+     * 3. finalSlots: [k1, k2, GROUPING_PREFIX_k1]
+     */
+    private List<Slot> genOutputSlots(
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:377]: output maybe not slot, but contains slot, and you need sort it, so `getUsedSortedSlots` is more reasonable. And primary method should move to the tailing of the class.

And this method should be simplified to
```java
private List<Slot> getUsedSortedSlots(List<Expression> groupByExpressions, List<NamedExpression> outputExpressions)
    Set <Slot> allSlots = ImmutableSet.builder()
        .addAll(groupByExpressions.stream()
            .flatMap(output -> output.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        .addAll(outputExpressions.stream()
            .flatMap(output -> output.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        ).build();

    Set<Slot> nonVirtualSlots = allSlots.stream()
        .filter(slot -> !(slot instanceof VirtualSlotReference)
        .collect(ImmutableSet.toSet());

    Set<Slot> virtualSlots = allSlots.stream()
        .filter(slot -> !(slot instanceof VirtualSlotReference)
        .collect(ImmutableSet.toSet());

    return ImmutableList.Builder()
        .addAll(nonVirtualSlots)
        .addAll(virtualSlots);
}
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 10]
```diff
@@ -257,6 +283,210 @@ public PlanFragment visitPhysicalAggregate(
...
+        nonVirtualSlots.addAll(nonVirtualOutput.stream()
+                .filter(e -> !nonVirtualSlots.contains(e))
+                .filter(SlotReference.class::isInstance)
+                .map(NamedExpression::toSlot)
+                .collect(Collectors.toSet()));
+        List<Slot> finalSlots = Lists.newArrayList();
+        finalSlots.addAll(nonVirtualSlots);
+        finalSlots.addAll(virtualSlots);
+        return finalSlots;
+    }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:400]: output maybe not slot, but contains slot, and you need sort it, so getUsedSortedSlots is more reasonable. And primary method should move to the tailing of the class.

And this method should be simplified to
```java
private List<Slot> getUsedSortedSlots(List<Expression> groupByExpressions, List<NamedExpression> outputExpressions)
    Set <Slot> allSlots = ImmutableSet.builder()
        .addAll(groupByExpressions.stream()
            .flatMap(groupExpr -> groupExpr.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        .addAll(outputExpressions.stream()
            .flatMap(outputExpr -> outputExpr.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        ).build();

    Map<Boolean, Set<Slot>> whetherIsVirtualSlots = groupBySlot.stream()
                .collect(Collectors.groupingBy(VirtualSlotReference.class::isInstance,
                        LinkedHashMap::new, ImmutableSet.toImmutableSet()));

    return ImmutableList.Builder()
        .addAll(whetherIsVirtualSlots.get(false)) // non virtual slots
        .addAll(whetherIsVirtualSlots.get(true)) // virtual slots
        .build();
}
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 11]
```diff
@@ -257,6 +283,210 @@ public PlanFragment visitPhysicalAggregate(
...
+        preExpressions.addAll(nonVirtualGroupByExpressions);
+        Map<Boolean, List<NamedExpression>> allOutput = outputExpressionList.stream()
+                .collect(Collectors.groupingBy(VirtualSlotReference.class::isInstance,
+                        LinkedHashMap::new, Collectors.toList()));
+        List<NamedExpression> nonVirtualOutput = allOutput.get(false);
+        nonVirtualOutput.stream()
+                .filter(e -> !preExpressions.contains(e))
+                .forEach(preExpressions::add);
+        return preExpressions;
+    }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:421]: ditto
```java
private List<Expression> genPreExpressions(
            List<Expression> nonVirtualGroupByExpressions, List<NamedExpression> outputExpressionList) {
    Set <Slot> allSlots = ImmutableSet.builder()
        .addAll(groupByExpressions.stream()
            .flatMap(groupExpr -> groupExpr.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        .addAll(outputExpressions.stream()
            .flatMap(outputExpr -> outputExpr.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        ).build();

    return allSlots.stream()
        .filter(slot -> !(slot instanceof VirtualSlotReference))
        .collect(ImmutableList.toImmutableList());
}
```

Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:421]: Look seems the used slots in aggregate/repeat will use lots of time, so you can extract all used slot, and extract slots according to your needs.

```java
Set<Slot> usedSlots = extractUsedSlots(groupByExpressions, outputExpressions);

private List<Slot> getUsedSortedSlots(Set<Slot> allSlot);
private List<Expression> getUsedNonVirtualSlots(Set<Slot> allSlot);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 12]
```diff
@@ -198,6 +220,10 @@ public PlanFragment visitPhysicalAggregate(
             slotList.addAll(groupSlotList);
             slotList.addAll(aggFunctionOutput);
             outputTupleDesc = generateTupleDesc(slotList, null, context);
+
+            if (needSetSlotToNullable(outputExpressionList)) {
+                setSlotNullable(outputTupleDesc, slotList);
+            }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:226]: if the invoke method is enough simple, please let the code reader understand the core logic without enter the method.
```suggestion
            if (containsGroupingScalarFunctioncOrVirtualSlot(outputExpressionList)) {
                setNullableToNonVirtualSlotReference(outputTupleDesc, slotList);
            }
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 13]
```diff
@@ -133,6 +142,26 @@ public PlanFragment translatePlan(PhysicalPlan physicalPlan, PlanTranslatorConte
         return rootFragment;
     }
 
+    private List<SlotReference> genGroupBySlotList(List<Expression> groupByExpressionList,
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:145]: rename to `collectGroupBySlots`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 14]
```diff
@@ -257,6 +283,210 @@ public PlanFragment visitPhysicalAggregate(
...
+         * eg: select sum(k2), grouping(k1) from t1 group by grouping sets((k1));
+         * finalSlots: k1, k2, GROUPING_ID(), GROUPING_PREFIX_k1
+         * prePressions: k1, k2
+         */
+        // create virtual tupleDesc
+        List<VirtualSlotReference> virtualSlotList = virtualGroupByExpressions.stream()
+                .filter(VirtualSlotReference.class::isInstance)
+                .map(VirtualSlotReference.class::cast)
+                .collect(Collectors.toList());
+        List<Slot> virtualSlots = Lists.newArrayList(virtualSlotList);
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:328]: this copy is useless.
```java
List<Slot> virtualSlotList = virtualGroupByExpressions.stream()
                .filter(VirtualSlotReference.class::isInstance)
                .map(Slot.class::cast)
                .collect(Collectors.toList());
```

**Review by 924060929 — COMMENTED** [review]
I think NormalizeXxx should refactor again

**Comment by morrySnow** [comment]
duplicated with #14430

**Pull request closed without merging**